Load and Save Sparse Matrix (arma::sp_mat) in MLpack using arma::coord_ascii

Hemal-Mamtora opened the pull request:
Aims at loading and saving a sparse matrix in mlpack.
Uses the `arma::coord_ascii`, `arma::arma_binary` formats and supports `.bin`, `.txt` and `.tsv` extensions

Cannot support `.csv` yet since `arma::coord_ascii` doesn't support csv.

Note: ARMA version: 8.400.0 (Entropy Bandit) does not support csv for sparse matrix.

birm commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -393,6 +393,38 @@ bool Load(const std::string& filename,
+  {
+    inplace_transpose(matrix);
+  }
+  
```
```suggestion

```

birm commented on src/mlpack/tests/load_save_test.cpp:
```diff
@@ -76,6 +76,42 @@ BOOST_AUTO_TEST_CASE(LoadCSVTest)
+    BOOST_REQUIRE_CLOSE(val, temp, 1e-5);
+  }
+  // Remove the file.
+  
```
```suggestion

```

birm commented on src/mlpack/tests/load_save_test.cpp:
```diff
@@ -76,6 +76,42 @@ BOOST_AUTO_TEST_CASE(LoadCSVTest)
+  arma::sp_mat::const_iterator it_end = test.end();
+
+  double temp = 0.0;
+  for(; it != it_end; ++it, temp += 0.1)
```
```suggestion
  for( ; it != it_end; ++it, temp += 0.1)
```

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -63,8 +63,8 @@ void TransposeTokens(std::vector<std::vector<std::string>> const &input,
-template<typename eT>
-bool inline inplace_transpose(arma::Mat<eT>& X)
+template <typename eT, template <typename> typename MatType>
+bool inline inplace_transpose(MatType<eT>& X)
```
Why not just remove the `eT` parameter and use `MatType` as the only template parameter?

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -393,6 +393,37 @@ bool Load(const std::string& filename,
+
+  if (transpose)
+  {
+    inplace_transpose(matrix);
```
Shouldn't we handle the output from `inplace_transpose()`?  (I suppose this would also apply to the other overloads of `Load()`.)

Hemal-Mamtora replied:
@rcurtin  
Thanks for the review!

Could you help me understand what is meant by "handling the output from `inplace_transpose()`" ?

rcurtin replied:
The function returns a `bool`---we're not handling that in any way.  I think that we should handle it. :+1:

rcurtin commented on src/mlpack/core/data/save_impl.hpp:
```diff
@@ -197,6 +197,45 @@ bool Save(const std::string& filename,
+          bool transpose)
+{
+  Timer::Start("saving_data");
+  arma::file_type saveType = arma::coord_ascii;
```
There are multiple types of sparse formats that can be loaded; I'd suggest that we use `Extension()` like for the dense matrices, and if the extension is not `.txt`, we don't infer that the type is `coord_ascii`.  Also, can we use `csv_ascii` when the type is `.csv`, and `arma_binary` when the type is `.bin`?

Hemal-Mamtora replied:
If I am correct,
The above suggestion would help answer this question: (raised in the description of the PR)

> What if an arma::mat contains a sparse matrix ? How to load that?
> Answer: If a sparse matrix is in dense format (in a csv or tsv having a lot of zeros), The **_user_** can load it using arma::mat(regular mlpack data::Load()) and then convert to arma::sp_mat

So please clarify this doubt:
input files for arma::sp_mat can be of 2 types:
1. Dense representation (having lots of zeros) which would be loaded with the help of arma::csv_ascii, arma::arma_ascii, arma::arma_binary etc.

2. arma::coord_ascii (of shape [n, 3]) which would load the sparse matrix using locations and values.

Are these types mutually exclusive ?
Or do I have to write some logic to determine either of the format types ?

Here, I understand the need of #2084 where manual type specification would have helped much. The user would simply specify arma::coord_ascii in the function call to load the file. 

Could you provide clarity on this ?

rcurtin replied:
Right, so there is some ambiguity in the CSV format.  The same Armadillo code, loading a sparse matrix with type `csv_ascii`, work for both of these CSV files:

("narrow" 3-column format)
```
1, 2, 3
1, 5, 6
7, 10, 11
```

("wide" dense format)
```
1, 2, 3, 4, 5
6, 7, 8, 0, 10
10, 0, 0, 6, 4
```

So, Armadillo does this format detection internally.  I think that we can just hand the file to Armadillo and let it do what it automatically does, using the file extension to auto-detect the format (`.csv` -> `csv_ascii`, `.txt` -> `coord_ascii`, `.bin` -> `arma_binary`).  If the format detection doesn't work correctly, then I think the user should specify a different file extension / file type, or try to load it as a dense matrix.  The user does have control over whether the matrix is loaded as dense or sparse simply via the type of the matrix they are loading.

Let me know what you think.  It's possible that I've overlooked something.

rcurtin commented on src/mlpack/tests/load_save_test.cpp:
```diff
@@ -76,6 +76,44 @@ BOOST_AUTO_TEST_CASE(LoadCSVTest)
+  for ( ; it != it_end; ++it, temp += 0.1)
+  {
+    double val = (*it);
+    BOOST_REQUIRE_CLOSE(val, temp, 1e-5);
```
Can we check that the rows and columns are coming back correctly too?  I know it's a little paranoid, but it's still probably good to check. :+1:

rcurtin commented on src/mlpack/core/data/load.cpp:
```diff
@@ -45,6 +45,36 @@ template bool Load<double>(const std::string&,
+                           const bool);
+
+template bool Load<float>(const std::string&,
+                           arma::SpMat<float>&,
+                           const bool,
+                           const bool);
```
```suggestion
                          arma::SpMat<float>&,
                          const bool,
                          const bool);
```

rcurtin commented on src/mlpack/core/data/load.hpp:
```diff
@@ -103,6 +109,11 @@ extern template bool Load<double>(const std::string&,
                                   const bool,
                                   const bool);
 
+extern template bool Load<double>(const std::string&,
+                                  arma::SpMat<double>&,
+                                  const bool,
+                                  const bool);
```
Shouldn't we add explicit instantiations for the other 5 types too?

Hemal-Mamtora commented:
@rcurtin 
For loading multiple sparse formats, I require some clarity and some assistance.

The way I see it:
There are 2 kinds of input formats possible:
1. a csv, or txt file having lots of zeros which actually makes it sparse but it is in dense format in the csv. (Type A)
2. a list of triplets (row_coord, column_coord, value), (Type B)

I have provided the support for Type B. And you are suggesting add support for type A too, correct ?

This means that I would have to write logic to distinguish between Type A input and Type B input. Any ideas on how to do that? The Naive Solution of if `shape == (3, N) `or `shape == (N, 3)` does not justify that the type is B.

The binary formats like arma::arma_binary would not cause problems, because the binary format would have the information between type A or type B. So not discussing that.

Manual type specification could be a solution.

rcurtin commented on src/mlpack/core/data/load.hpp:
```diff
@@ -103,6 +109,31 @@ extern template bool Load<double>(const std::string&,
+extern template bool Load<unsigned int>(const std::string&,
+                                        arma::SpMat<unsigned int>&,
+                                        const bool,
+                                        const bool);
```
Don't forget the `int` overload. :)

birm replied:
seems to be addressed in 8cac6f6

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -63,17 +63,24 @@ void TransposeTokens(std::vector<std::vector<std::string>> const &input,
+      Log::Warn << "\nTranspose Operation Failed.\n"
+          "Exception: " << e.what() << std::endl;
+    
+    return false;
```
Thanks!  This at least does something better than before.  Maybe later we should revisit it... it doesn't look like it actually attempts to do the transpose in-place (maybe Armadillo does that internally now, I'm not sure).  So in my view the function is a candidate for removal or overhaul, but, we can leave that for another time. :)

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -290,7 +297,9 @@ bool Load(const std::string& filename,
   if (transpose)
   {
-    inplace_transpose(matrix);
+    bool is_transposed = inplace_transpose(matrix, fatal);
```
```suggestion
    bool isTransposed = inplace_transpose(matrix, fatal);
```
Simple style guide fix; we should name variables in camelcase. :)

Hemal-Mamtora commented:
@rcurtin 

So, it seems that the implementation would be too similar to that of Load() for `arma::Mat`. So I was thinking of having a common implementation for `arma::Mat` and `arma::SpMat` by having a template parameter `MatType<eT>` or `MatType` which should have value of `arma::Mat<eT>` by default. 

I would like to try this.

Note: I'll keep the existing Load() function as it is required for backward compatibility.

But if a common load function comes into picture, the lines of code would get reduced. I am not sure about the speed of execution though.

I would like to try this; could you give me some head start? I am getting confused in template parameters.
Which one to use:
1. template \<typename MatType\>
2. template \<typename eT, template \<typename\> typename MatType\>

for 
```
bool Load(const std::string& filename,
          MatType<eT>& matrix,
          const bool fatal = false,
          const bool transpose = true);
```

Hemal-Mamtora commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -393,6 +406,39 @@ bool Load(const std::string& filename,
+{
+  bool success;
+  Timer::Start("loading_data");
+  success = matrix.load(filename, arma::coord_ascii);
```
@rcurtin 

I tried using arma::csv_ascii instead of arma::coord_ascii, here
Both the new tests which I wrote namely : SaveSparseCSVTest, and LoadSparseCSVTest have failed.
It seems that arma::coord_ascii is the only method to load a narrow matrix.

(ARMA version: 8.400.0 (Entropy Bandit))

rcurtin replied:
Ah, sorry, I think you are right here.  I misread the output of my experiments.  So then in this case I would say that we should use `coord_ascii` for a `.csv` file.  Later, we can figure out how to let the user specify the type manually.

rcurtin commented:
> @rcurtin
> 
> So, it seems that the implementation would be too similar to that of Load() for `arma::Mat`. So I was thinking of having a common implementation for `arma::Mat` and `arma::SpMat` by having a template parameter `MatType<eT>` or `MatType` which should have value of `arma::Mat<eT>` by default.
> 
> I would like to try this.
> 
> Note: I'll keep the existing Load() function as it is required for backward compatibility.
> 
> But if a common load function comes into picture, the lines of code would get reduced. I am not sure about the speed of execution though.
> 
> I would like to try this; could you give me some head start? I am getting confused in template parameters.
> Which one to use:
> 
>     1. template <typename MatType>
> 
>     2. template <typename eT, template <typename> typename MatType>
> 
> 
> for
> 
> ```
> bool Load(const std::string& filename,
>           MatType<eT>& matrix,
>           const bool fatal = false,
>           const bool transpose = true);
> ```

I would suggest against this, honestly.  If you are going to do it, you'll have to templatize the code that figures out the type from the extension, since the set of allowable types is different for sparse and dense matrices.  If you did do it like that, you would want to templatize with `template<typename MatType>`, but then you'd also want to use SFINAE to make sure that `MatType` is only ever an `arma::Mat<eT>` or `arma::SpMat<eT>`.

The implementation of `data::Load(..., SpMat<eT>&, ...)` should be pretty short, I think.

Hemal-Mamtora commented:
@rcurtin 

I've added functionality to load and save sparse matrix in .txt, .csv, .tsv files of type : arma::coord_ascii

There was earlier a bug in loading the .csv file which I've fixed but the fix is not clean.
For arma::coord_ascii, the format expects data without commas. If the data is with commas, it does not load correctly.

So to load a n*3 csv, I had to remove the commas and then loaded them. To do that, I created a new temp file without commas and loaded  it from there. 

If there is a better way to do that, please let me know. Perhaps some stream_processing.

I guess, the memory fails are due to that only.

Please guide me on how to proceed further.

rcurtin commented on src/mlpack/core/data/save_impl.hpp:
```diff
@@ -197,6 +197,194 @@ bool Save(const std::string& filename,
+  // Finally return success.
+  return true;
+}
+  // Timer::Start("saving_data");
+  // arma::file_type saveType = arma::coord_ascii;
+  // std::string stringType = "CSV data";
+  // arma::SpMat<eT> tmp;
+
+  // if (transpose)
+  //   tmp = trans(matrix);
+  // else
+  //   tmp = matrix;
+
+  // // Try to save the file.
+  // Log::Info << "Saving " << stringType << " to '" << filename << "'."
+  //     << std::endl;
+
+  // const bool success = tmp.quiet_save(filename, saveType);
+
+  // if (!success)
+  //   {
+  //     Timer::Stop("saving_data");
+  //     if (fatal)
+  //       Log::Fatal << "Save to '" << filename << "' failed." << std::endl;
+  //     else
+  //       Log::Warn << "Save to '" << filename << "' failed." << std::endl;
+
+  //     return false;
+  //   }
+
+  // Timer::Stop("saving_data");
+  // // Finally return success.
+  // return true;
+// }
```
It looks like this is not needed anymore?

rcurtin commented on src/mlpack/core/data/save_impl.hpp:
```diff
@@ -197,6 +197,194 @@ bool Save(const std::string& filename,
+    saveType = arma::arma_binary;
+    stringType = "Armadillo binary formatted data";
+  }
+  else if (extension == "pgm")
+  {
+    saveType = arma::pgm_binary;
+    stringType = "PGM data";
+  }
+  else if (extension == "h5" || extension == "hdf5" || extension == "hdf" ||
+           extension == "he5")
```
I'm not sure why you've included these formats?  Armadillo sparse matrices can't be saved in this format as far as I know.

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -393,6 +406,187 @@ bool Load(const std::string& filename,
+    std::istreambuf_iterator<char> in_itr(stream);
+    std::ostreambuf_iterator<char> out_itr(csvstream);
+
+    // replacing comma with a space and storing to csvstream
+    std::replace_copy(in_itr,
+                      std::istreambuf_iterator<char>(),
+                      out_itr,
+                      ',',
+                      ' ');
```
You're right, `coord_ascii` doesn't work when there are commas.  So, therefore, I think that we should only support `.txt` and remove this code (sorry about that, I know that you put time into it).  If Armadillo later adds support for commas, then we can add `.csv` here too.

Hemal-Mamtora replied:
So could we use CSV to support broad sparse matrices ?

Hemal-Mamtora replied:
No, we can't use arma::csv_ascii as of now.

(I checked the current arma version ) (ARMA version: 8.400.0 (Entropy Bandit))

Currently, only arma::coord_ascii and arma::arma_binary are supported in SpMat_meat.hpp load() function. 

So you are right, only txt should be supported, 
But, tsv files having \\t can also be supported. So keeping txt, tsv and bin.

rcurtin replied:
Yep, I agree with your assessment here.  Maybe later we will be able to add formats but this is good for now. :+1:

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -290,7 +297,9 @@ bool Load(const std::string& filename,
   if (transpose)
   {
-    inplace_transpose(matrix);
+    bool isTransposed = inplace_transpose(matrix, fatal);
+    if (!isTransposed)
+      return false;
```
Maybe just `return inplace_transpose(matrix, fatal)`?

rcurtin commented on src/mlpack/core/data/save.hpp:
```diff
@@ -59,6 +59,12 @@ bool Save(const std::string& filename,
+bool Save(const std::string& filename,
+          const arma::SpMat<eT>& matrix,
+          const bool fatal = false,
+          bool transpose = true);
```
Can you add comprehensive documentation to this?  This is a user-facing function, so it's really important that it's documented very clearly. :+1:

rcurtin replied:
Looks good, thank you. :)

rcurtin commented on src/mlpack/core/data/load.hpp:
```diff
@@ -65,6 +65,12 @@ bool Load(const std::string& filename,
+bool Load(const std::string& filename,
+          arma::SpMat<eT>& matrix,
+          const bool fatal = false,
+          const bool transpose = true);
```
Can you add comprehensive documentation to this? This is a user-facing function, so it's really important that it's documented very clearly. :+1:

rcurtin replied:
Awesome, this looks great.  Thank you! :+1:

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -393,6 +402,130 @@ bool Load(const std::string& filename,
+    }
+    else // We can only assume it's raw binary.
+    {
+      stringType = "raw binary formatted data";
```
```suggestion
      stringType = "raw binary formatted data for sparse matrix";
```
(just for consistency)

Hemal-Mamtora commented:
@rcurtin , I'm glad to be able to contribute !

Regarding the history.md, could you do it during the merge process ?

This is because, there had been an issue with my system while upgrading it to Ubuntu 20.04

So it would take me a while to setup mlpack again.

rcurtin commented:
@Hemal-Mamtora no problem, I can handle it. :+1: Thanks again!
